opsimulator/l2tol2messagestore.go

```diff
@@ -0,0 +1,191 @@
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	entry, exists := s.entryByHash[msgHash]
+	if !exists {
+		return nil, fmt.Errorf("message not found")
+	}
+	return entry, nil
+}
+
+type UpdaterFunc func(lifecycle *L2ToL2MessageLifecycle) (*L2ToL2MessageLifecycle, error)
+
+func (s *L2ToL2MessageStore) UpdateLifecycle(msgHash common.Hash, updater UpdaterFunc) (*L2ToL2MessageStoreEntry, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	entry, exists := s.entryByHash[msgHash]
+	if !exists {
+		return nil, fmt.Errorf("message not found")
+	}
+
+	newLifecycle, err := updater(entry.lifecycle)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update lifecycle: %w", err)
+	}
+
+	newEntry := &L2ToL2MessageStoreEntry{
+		message:   entry.message,
+		lifecycle: newLifecycle,
+	}
+
+	s.entryByHash[msgHash] = newEntry
+
+	return newEntry, nil
+}
+
+type L2ToL2MessageStoreManager struct {
+	store *L2ToL2MessageStore
+}
+
+func NewL2ToL2MessageStoreManager() (*L2ToL2MessageStoreManager, error) {
+	return &L2ToL2MessageStoreManager{
+		store: NewL2ToL2MessageStore(),
+	}, nil
+}
+
+func (m *L2ToL2MessageStoreManager) HandleSentEvent(log *types.Log) error {
+	msg, err := NewL2ToL2MessageFromSentMessageEventData(log.Data)
+	if err != nil {
+		return fmt.Errorf("failed to create L2ToL2Message: %w", err)
+	}
+
+	msgHash, err := msg.Hash()
+	if err != nil {
+		return fmt.Errorf("failed to calculate message hash: %w", err)
+	}
+
+	lifecycle := &L2ToL2MessageLifecycle{
+		SentTxHash: log.TxHash,
+	}
+
+	entry := &L2ToL2MessageStoreEntry{
+		message:   msg,
+		lifecycle: lifecycle,
+	}
+
+	return m.store.Set(msgHash, entry)
+}
+
+func (m *L2ToL2MessageStoreManager) HandleRelayedEvent(log *types.Log) error {
+	if log.Topics[0] != bindings.L2ToL2CrossDomainMessengerParsedABI.Events["RelayedMessage"].ID {
+		return fmt.Errorf("unexpected event type")
+	}
+
+	msgHash := log.Topics[1]
+
+	_, err := m.store.UpdateLifecycle(msgHash, func(lifecycle *L2ToL2MessageLifecycle) (*L2ToL2MessageLifecycle, error) {
+		if lifecycle.RelayedTxHash != (common.Hash{}) {
+			return nil, fmt.Errorf("message already relayed")
+		}
+
+		return lifecycle.WithRelayedTxHash(log.TxHash), nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to update lifecycle for relayed event: %w", err)
+	}
+
+	return nil
+}
+
+func (m *L2ToL2MessageStoreManager) HandleFailedRelayedEvent(log *types.Log) error {
+	if log.Topics[0] != bindings.L2ToL2CrossDomainMessengerParsedABI.Events["FailedRelayedMessage"].ID {
+		return fmt.Errorf("unexpected event type")
+	}
+
+	msgHash := log.Topics[1]
+
+	_, err := m.store.UpdateLifecycle(msgHash, func(lifecycle *L2ToL2MessageLifecycle) (*L2ToL2MessageLifecycle, error) {
+		if lifecycle.RelayedTxHash != (common.Hash{}) {
+			return nil, fmt.Errorf("message already relayed")
+		}
+
+		return lifecycle.WithFailedTxHash(log.TxHash), nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to update lifecycle for failed relayed event: %w", err)
+	}
```

Comment:
nit on the extra line on 181 not being needed
```suggestion
	if _, err := m.store.UpdateLifecycle(msgHash, func(lifecycle *L2ToL2MessageLifecycle) (*L2ToL2MessageLifecycle, error) {
		if lifecycle.RelayedTxHash != (common.Hash{}) {
			return nil, fmt.Errorf("message already relayed")
		}

		return lifecycle.WithFailedTxHash(log.TxHash), nil
	}); err != nil {
		return fmt.Errorf("failed to update lifecycle for failed relayed event: %w", err)
	}
```